### PR TITLE
Decoupled Tandem Slice Projection: separate routing matrix for tandem samples

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

In Transolver, each mesh node is routed to a mixture of physics slices via a shared projection matrix `in_project_slice`. For **tandem foil** configurations, the flow topology is fundamentally different from single-foil cases: inter-foil pressure channels, wake interactions, and aft-foil leading-edge stagnation create routing patterns that don't exist in single-foil physics.

Currently, `in_project_slice` must simultaneously encode both single-foil and tandem routing vocabularies in the same weight matrix. The `--adaln_decouple` flag (already implemented in `Physics_Attention_Irregular_Mesh`, line 165-167) enables a **separate orthogonally-initialized** `in_project_slice_tandem` matrix used exclusively for tandem samples. This lets the model specialize tandem slices for tandem-specific physics — inter-foil channel nodes, wake regions, aft-foil leading edge — without contaminating the single-foil routing.

**Key insight:** PCGrad, gap/stagger spatial bias, and the TE coordinate frame all improved results by better separating tandem vs. single-foil physics. Decoupled slice routing attacks the same axis but at the most fundamental level — the slice attention routing itself.

**This is a zero-code-change experiment — add one flag only.**

Code path:
```python
# Physics_Attention_Irregular_Mesh.__init__ (line 165-168)
if decouple_slice:
    self.in_project_slice_tandem = nn.Linear(dim_head, slice_num)
    torch.nn.init.orthogonal_(self.in_project_slice_tandem.weight)
# Forward (line 212-214)
if self.decouple_slice and tandem_mask is not None:
    tan_logits = self.in_project_slice_tandem(x_mid) / temp
```

## Instructions

**No code changes needed.** Add `--adaln_decouple` to the baseline command.

Run both seeds:
```bash
cd cfd_tandemfoil && python train.py --agent frieren \
  --wandb_name "frieren/adaln-decouple-s42" \
  --wandb_group "frieren/adaln-decouple-tandem" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --adaln_decouple \
  --seed 42

cd cfd_tandemfoil && python train.py --agent frieren \
  --wandb_name "frieren/adaln-decouple-s73" \
  --wandb_group "frieren/adaln-decouple-tandem" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --adaln_decouple \
  --seed 73
```

**Expected VRAM overhead:** Minimal — one extra `nn.Linear(dim_head, slice_num)` ≈ negligible parameters.

**When to abort:** If after 30 epochs p_tan > 30.0, stop seed 42 early and report.

## Baseline (PR #2213, 2-seed avg)

| Metric | Value | Target to beat |
|--------|-------|----------------|
| p_in | 11.979 | < 11.98 |
| p_oodc | 7.643 | < 7.65 |
| p_tan | 28.341 | < 28.34 |
| p_re | 6.300 | < 6.30 |

Baseline W&B runs: `hgml7i2r` (s42), `qic03vrg` (s73)

Reproduce baseline:
```bash
cd cfd_tandemfoil && python train.py --agent frieren --wandb_name "frieren/baseline-wake-deficit" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT frieren:

## Results

### adaln_decouple, 2-seed

| Metric | Baseline (2-seed avg) | Decouple (seed 42) | Decouple (seed 73) | Decouple (2-seed avg) | Δ vs baseline |
|--------|----------------------|---------------------|---------------------|----------------------|---------------|
| **p_in** | **11.979** | 14.007 | 12.761 | **13.384** | **+11.7%** ❌ |
| **p_oodc** | **7.643** | 8.668 | 8.552 | **8.610** | **+12.7%** ❌ |
| **p_tan** | **28.341** | 28.646 | 28.910 | **28.778** | **+1.5%** ❌ |
| **p_re** | **6.300** | 6.985 | 6.911 | **6.948** | **+10.3%** ❌ |

### Run details
- **Seed 42**: W&B run `wc67srhn`, best epoch 139/500, peak VRAM 50.6GB, 180 min
- **Seed 73**: W&B run `mhi0x3ve`, best epoch 140/500, peak VRAM 50.5GB, 180 min
- Both completed ~147 epochs (EMA from epoch 140)

### Command used
```bash
cd cfd_tandemfoil && python train.py --agent frieren \
  --wandb_name "frieren/adaln-decouple-s{42,73}" \
  --wandb_group "frieren/adaln-decouple-tandem" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --adaln_decouple \
  --seed {42,73}
```

### What happened

**Negative result.** Decoupled tandem slice routing degraded all four pressure metrics, most significantly p_in (+11.7%), p_oodc (+12.7%), and p_re (+10.3%).

**Why it likely failed:**

1. **Insufficient tandem training signal:** The tandem-specific `in_project_slice_tandem` matrix sees only tandem samples (~30% of training data), meaning it gets far fewer gradient updates than the shared routing matrix. With a batch size of 4 and `tandem_ramp` reducing tandem contribution in early epochs, the tandem routing head is severely under-trained.

2. **Orthogonal initialization diverges from learned routing:** The tandem routing starts from orthogonal init while the shared routing has learned meaningful physics-aware slice assignments. The tandem head must learn from scratch how to route flow features, lacking the transfer from single-foil training that the shared routing benefits from.

3. **Existing mechanisms already handle tandem/single separation:** `domain_layernorm` (separate LayerNorm per domain), `domain_velhead` (separate velocity heads), and `pcgrad_3way` (separate gradient optimization paths) already provide effective tandem/single separation. Adding another decoupling axis at the slice routing level fragments the representation rather than helping.

4. **Seed 42 especially bad on p_in (+17%):** High seed-to-seed variance (14.0 vs 12.8) suggests the tandem routing head is unstable — further evidence of under-training.

### Suggested follow-ups
- If this idea is revisited, try initializing `in_project_slice_tandem` from the trained shared weights (copy init) rather than orthogonal init — this gives tandem routing a head start
- A warm-start approach (shared routing first, then split after epoch X) might work better
- The large regression on p_in and p_oodc suggests the tandem routing interfered with single-foil representation learning, possibly through gradient flow

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2258 because ADVISOR: Closing as a dead end. All four metrics regressed significantly vs baseline: p_in +11.7% (13.384 vs 11.979), p_oodc +12.7% (8.610 vs 7.643), p_tan +1.5% (28.778 vs 28.341), p_re +10.3% (6.948 vs 6.300). Your analysis is exactly right — the orthogonal-initialized tandem routing head is severely undertrained on ~30% tandem samples, the existing domain_layernorm/domain_velhead/pcgrad_3way already provide effective tandem/single separation, and the seed-to-seed variance on p_in (14.0 vs 12.8 = 9% spread) confirms routing instability. If revisited: warm-start by copying the shared weights into in_project_slice_tandem at init, then fine-tune from there. But the existing separation mechanisms are likely sufficient..
